### PR TITLE
0.5 - Feature scan endpoint directory

### PR DESCRIPTION
**Pull request by peteraba**
https://github.com/gernest/apidemic/issues/3

PR https://github.com/gernest/apidemic/pull/8 recreated.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -158,6 +158,9 @@ There's also a 1% chance for the server to claim to be a [Teapot](https://tools.
 
 **Note**: JSON keys must be strings, providing your response codes as integers will not work!
 
+### Registering endpoints automatically
+The service will scan the `./endpoints/` directory relative to the currenct working directory on startup. You can change the directory to be scanned by adding a flag `--endpoint-dir=ENDPOINT_DIR`.
```
Comment (gernest) [inline README.md]: for relative paths `endpoints/` will do, no need for `./endpoints/`

File: helper.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package apidemic
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+)
+
+func JsonRequest(method string, path string, body interface{}) *http.Request {
```
Comment (gernest) [inline helper.go]: Do you plan to use this anywhere else apart from tests?

This only works in tests so no need to export the name or put it in not `*_test.go` file. And if in case you want to use this in something else you better add godoc comment since it is an exported API to say what it does.

File: cmd/apidemic/main.go [hunk 3]
```diff
@@ -38,8 +50,66 @@ func main() {
 					Value:  3000,
 					EnvVar: "PORT",
 				},
+				cli.StringFlag{
+					Name:   "endpoint-dir",
+					Usage:  "Directory to scan for endpoint",
+					Value:  "./endpoints/",
```
Comment (gernest) [inline cmd/apidemic/main.go]: `endpoints/`

File: cmd/apidemic/main.go [hunk 4]
```diff
@@ -1,18 +1,30 @@
...
 	"github.com/gernest/apidemic"
+	"github.com/gorilla/mux"
 )
 
 func server(ctx *cli.Context) {
 	port := ctx.Int("port")
+	endpointDir := ctx.String("endpoint-dir")
 	s := apidemic.NewServer()
 
+	err := addEndpoints(s, endpointDir)
```
Comment (gernest) [inline cmd/apidemic/main.go]: The endpoint dir should be optional. Not everyone wants to load stuff from a dir. I think the workflow should be like this.

 - If the user provided a path to the endoints then use it to lad endpoint and exit if any error occured.

- If the use didn't provide an endpoint , try to check if `endpoints/` is there if not continue execution ignore the loading . In case the dir is there ☝️  see above bullet but use `endpoints/` instead.

- Else just ignore the loading

File: cmd/apidemic/main.go [hunk 5]
```diff
@@ -38,8 +50,66 @@ func main() {
...
+		registerPayload, err = getRegisterPayload(filePath)
+		if err != nil {
+			return err
+		}
+
+		w := httptest.NewRecorder()
+		req := apidemic.JsonRequest("POST", "/register", registerPayload)
+		s.ServeHTTP(w, req)
+
+		log.Printf("%s is registered\n", filePath)
```
Comment (gernest) [inline cmd/apidemic/main.go]: No need, for this. It just adds noise.

File: cmd/apidemic/main.go [hunk 6]
```diff
@@ -38,8 +50,66 @@ func main() {
...
+
+		if w.Code != http.StatusOK {
+			return errors.New("registering " + filePath + " failed")
```
Comment (gernest) [inline cmd/apidemic/main.go]: use `fmt.Errorf`

**Review by gernest — CHANGES_REQUESTED** [review]
Please check the comment. I will take another look after the comments have been addressed.

File: README.md (hunk 1 above)
Comment (peteraba) [reply README.md]: it's true and I was considering it, but I feel ./endpoints is a bit more expressive. I'll change it but I'm not sure this will be an improvement.

File: cmd/apidemic/main.go (hunk 5 above)
Comment (peteraba) [reply cmd/apidemic/main.go]: Actually I found this very useful and only adds noice at the beginning. Maybe I can add a verbose flag to see this log?

File: cmd/apidemic/main.go (hunk 3 above)
Comment (peteraba) [reply cmd/apidemic/main.go]: :ok:

File: README.md (hunk 1 above)
Comment (peteraba) [reply README.md]: :ok:

File: cmd/apidemic/main.go (hunk 4 above)
Comment (peteraba) [reply cmd/apidemic/main.go]: :ok:

File: helper.go (hunk 2 above)
Comment (peteraba) [reply helper.go]: This is already used when reading the endpoints. https://github.com/gernest/apidemic/pull/12/files#diff-510e51effef9a268bfb7ca60921904daR88

**Comment by peteraba** [comment]
@gernest please check my responses before I fix the remaining issues.

File: cmd/apidemic/main.go (hunk 5 above)
Comment (gernest) [reply cmd/apidemic/main.go]: I see, you can leave this for now.

File: helper.go (hunk 2 above)
Comment (gernest) [reply helper.go]: Okay, just add comments then

**Comment by peteraba** [comment]
@gernest please check my changes.

**Comment by peteraba** [comment]
@gernest any chance you could check this?

**Comment by gernest** [comment]
Hey sorry. I am checking it right now.

File: cmd/apidemic/main.go [hunk 7]
```diff
@@ -1,18 +1,35 @@
...
 )
 
 func server(ctx *cli.Context) {
 	port := ctx.Int("port")
+	endpointDir := ctx.String("endpoint-dir")
 	s := apidemic.NewServer()
 
+	_, err := os.Stat(endpointDir)
+	if err != nil && endpointDir != "endpoints/" {
+		fmt.Errorf("Endpoint not found: %s", endpointDir)
```
Comment (gernest) [inline cmd/apidemic/main.go]: no caps in error message. Should be `endpoint not found`

File: cmd/apidemic/main.go [hunk 8]
```diff
@@ -38,8 +55,66 @@ func main() {
...
+	if err != nil {
+		return nil
+	}
+
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+
+		filePath := endpointDir + file.Name()
```
Comment (gernest) [inline cmd/apidemic/main.go]: use `filepath.Join(endpointDir ,file.Name())` for cross platform use

**Review by gernest — CHANGES_REQUESTED** [review]
Okay looks nice. I left one comment. I noticed there are no new tests for the  `addEndpoints` function.

**Comment by peteraba** [comment]
@gernest I fixed the issues you pointed out in the code change.

I did not add tests for the `addEndpoints` function as there's currently none for the `main` package. If you really want to see those added, please give me some pointers on what exactly should be tested.

**Comment by peteraba** [comment]
@gernest would be cool to have these PRs closed as soon as possible. I won't be opening more anytime soon I think :wink:

**Comment by peteraba** [comment]
@gernest ?

**Comment by gernest** [comment]
@peteraba I'm sorry. I am very sick right now. I will take another look the moment I recover and feel better.

**Comment by peteraba** [comment]
@gernest understood, get well soon!
